js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,323 @@
+        // Update actions/buttons
+        this.actions = [];
+
+        if (this._buttonBox) {
+            this._buttonBox.destroy();
+            this._buttonBox = null;
+        }
+
+        let buttons = params.buttons;
+        if (buttons) {
+            buttons.deep_unpack().forEach(Lang.bind(this, function(button) {
+                this.addAction(button.label.unpack(),
+                               Lang.bind(this, this._onButtonClicked, button));
+            }));
+        }
+
+        // Update icon
+        let newGicon = null;
+        let gicon = params.icon;
+        if (gicon)
+            newGicon = this._getClubhouseIcon(Gio.icon_deserialize(gicon));
+
+        // Update time
+        let timestamp = params.timestamp;
+        let time = timestamp ? GLib.DateTime.new_from_unix_local(timestamp) : null;
+
+        this.update('', params.body.unpack(), {gicon: newGicon, datetime: time});
+    },
+});
+
+var ClubhouseNotificationSource = new Lang.Class({
+    Name: 'ClubhouseNotificationSource',
+    Extends: NotificationDaemon.GtkNotificationDaemonAppSource,
+
+    _init: function(appId) {
+        this.parent(appId);
+        this._notification = null;
+    },
+
+    addNotification: function(notificationId, notificationParams, showBanner) {
+        if (this._notification == null)
+            this._notification = new ClubhouseNotification(this, notificationParams);
+        else
+            this._notification.updateFromParams(notificationParams);
+
+        this.notify(this._notification);
+    },
+});
+
+var ClubhouseComponent = new Lang.Class({
+    Name: 'ClubhouseComponent',
+
+    _init: function() {
+        this._useClubhouse = false;
+
+        let eosImageVersion = Shell.util_get_eos_image_version();
+        if (eosImageVersion == null || !this._imageUsesClubhouse(eosImageVersion))
+            return;
+
+        this._useClubhouse = true;
+        this._banner = null;
+        this.actor = null;
+        this.source = null;
+        this._clubhouseProxy = null;
+        this._clubhouseProxyHandler = 0;
+        this._oldAddNotificationFuncPrototype = null;
+    },
+
+    enable: function() {
+        if (!this._useClubhouse) {
+            log('Cannot enable Clubhouse in this image version');
+            return;
+        }
+
+        log('Enabling Clubhouse component');
+
+        this._clubhouseProxy = new Gio.DBusProxy.new_sync(sessionBus,
+                                                          Gio.DBusProxyFlags.DO_NOT_AUTO_START_AT_CONSTRUCTION,
+                                                          null,
+                                                          CLUBHOUSE_ID,
+                                                          CLUBHOUSE_DBUS_OBJ_PATH,
+                                                          CLUBHOUSE_ID,
+                                                          null);
+        this._clubhouseProxy.connect('notify::g-name-owner', Lang.bind(this, function() {
```

Comment:
We should use arrow functions instead of `Lang.bind` in new code.

Reply:
I will have to check how that's done.